### DISCUSSION TITLE
Adding a rudimentary search/filter based on a menu list.

### DISCUSSION THREAD

**Discussion by aaronstezycki** [alpinejs/alpine#3221] [1. Help] [answered]
Hi, so I've got a menu list of items already generated by a static site builder that I wish to filter. This doesn't seem too difficult, but as I'm a learning JS developer (UX/UI Designer mostly) I need some help getting this to work.

I'd like to create something similar to [this search](https://www.youtube.com/watch?v=lwFejyumY2Y), however this technique uses templates (which I don't need).

Any help with the following example would be appreciated!

Proposed example:

```html

<div x-data="
   items: querySelector('[data-tag]').split(","),
   search: "",
   get filteredItems() {
       // This is where I get stuck
       return this.items.filter(
           item => item.startsWith(this.search)
       )
   }
">
    <input type="search" x-model="search" />

    <ul>
        /* Set [hidden] on each item if the search term matches the data tag */
        <li data-tag="a" :hidden="item ? search">Menu Item</li>
        <li data-tag="b" :hidden="item ? search">Menu Item</li>
        <li data-tag="c" :hidden="item ? search">
            Menu Drop >
            <ul data-tag="c" :hidden="item ? search">
                <li data-tag="c, e" :hidden="item ? search">Menu Item</li>
                <li data-tag="c, d" :hidden="item ? search">Menu Item</li>
            </ul>
        </li>
        <li data="e">Menu Item</li>
    </ul>
</div>
```

As a plus, I reckon this template could be hugely popular as an Alpine JS Component. Thoughts?

**Comment by SimoTod** [comment]
If you don't want to use a template, your code will end up being quite static but you don't need items and filteredItems, you just use `:hidden="$el.dataset.tag != search"`

**Comment by aaronstezycki** [comment]
@SimoTod Yeah I understand the content will be static, there's a number of things going on on the page where it just 'feels safer' to use less javascript to do stuff. Although I'm loving Alpine... its very powerful!

The page uses barba.js for animations, so for the time being it's just easier to go static, than to put more work into getting alpine and barba.js to talk to each other.

Also, it's using 11ty to process the pages, so that makes it slightly easier.

As a side note, if items inside the `<templates>` also have further alpine directives do those continue to work by default or is there extra 'magic' needed to make that happen?

**Reply by ekwoka** [reply]
When the new elements get added to the page their directives and such are all evaluated properly.

It's a bit different process when the element is added by the x-for and x-if directives vs some other thing injecting the elements, but it all *just works*.

**Reply by aaronstezycki** [reply]
Thanks for that, I'll certainly have to try out `<templates>`, I know 11ty can generate a navigation json based on pages that could feed into an Alpine template component. Would be interesting to see how that comes out.

**Comment by aaronstezycki** [comment, marked as answer]
So I've ended up with something a little more in-depth, and processed the search input and items a bit more.
I use alpine to show/hide the nested `<ul>` nodes , but would be nice to show *everything* when the search has been used.

I've tried `<ul x-show="isOpen || show_item($el)">`, however that doesn't seem to work and just breaks the show/hide on the nested levels. What am I missing here?

```
<div class="wrapper">
    <div x-data="{
        search: '',
        show_item(el){
            return this.search === '' || el.textContent.toLowerCase().includes(this.search.toLowerCase());
         }
     }">
         <input class="search-input" type="search" placeholder="Filter: A B C D" x-model="search">
         <p>Search: <span x-text="search"></span></p>
          
         <ul class="menu">
             <li x-show="show_item($el)">Avatars</li>
             <li x-show="show_item($el)">Badges</li>
             <li x-show="show_item($el)">Buttons</li>
             <li x-data="{ isOpen: false }" x-show="show_item($el)">
                 <button @click="isOpen = ! isOpen">Forms</button>
                 <ul x-show="isOpen || show_item($el)">
                     <li x-show="show_item($el)">Fields</li>
                     <li x-show="show_item($el)">Inputs</li>
                     <li x-show="show_item($el)">Validation</li>
                 </ul>
             </li>
             <li x-show="show_item($el)">Icons</li>
             <li x-show="show_item($el)">Images</li>
             <li x-show="show_item($el)">Lists</li>
         </ul>
     </div>
</div>
```

**Reply by aaronstezycki** [reply]
Solved it. This works rather well now.

```
<div x-data="{
  search: '',
  show_item(el){
    return this.search === '' || el.textContent.toLowerCase().includes(this.search.toLowerCase());
  }
  }">
    <input class="search-input" type="search" placeholder="Filter: Icons" x-model="search">
    <p x-show="search != ''">Searching by: <span x-text="search"></span></p>

  <p><b>Menu:</b></p>
  <ul class="menu">
    <li x-show="show_item($el)">Avatars</li>
    <li x-show="show_item($el)">Badges</li>
    <li x-show="show_item($el)">Buttons</li>
    <li x-show="show_item($el)">Cards</li>
    <li x-show="show_item($el)">Dropdown</li>
    <li x-data="{ isOpen: false }" x-show="show_item($el)">
      <button @click="isOpen = ! isOpen">Forms</button>
      <ul x-show="isOpen || search != ''">
        <li x-show="show_item($el)">Fields</li>
        <li x-show="show_item($el)">Inputs</li>
        <li x-show="show_item($el)">Validation</li>
      </ul>
    </li>
    <li x-show="show_item($el)">Icons</li>
    <li x-show="show_item($el)">Images</li>
    <li x-show="show_item($el)">Lists</li>
    <li x-data="{ isOpen: false }" x-show="show_item($el)">
      <button @click="isOpen = ! isOpen">Navigation</button>
      <ul x-show="isOpen || search != ''">
        <li x-show="show_item($el)">Adaptive Navigation</li>
        <li x-show="show_item($el)">Navbar</li>
      </ul>
    </li>
  </ul>
  </div>
</div>

```
